src/daft-minhash/src/windowed.rs

```diff
@@ -0,0 +1,319 @@
+struct WindowedWords<'a> {
+    s: &'a str,
+    word_starts: Vec<usize>, // Vec of start indices for each word
+    window_size: usize,
+    current: usize, // Current starting word index for the window
+}
+
+pub trait WindowedWordsExt<'a> {
+    fn windowed_words(&'a self, window_size: usize) -> impl Iterator<Item = &'a str>;
+}
+
+impl<'a> WindowedWordsExt<'a> for str {
+    fn windowed_words(&'a self, window_size: usize) -> impl Iterator<Item = &'a Self> {
+        WindowedWords::new(self, window_size)
+    }
+}
+
+impl<'a> WindowedWords<'a> {
+    /// Creates a new `WindowedWords` iterator.
+    ///
+    /// # Arguments
+    ///
+    /// * `s` - The input string slice.
+    /// * `window_size` - The number of words in each window.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// let s = "The quick brown fox";
+    /// let iter = WindowedWords::new(s, 2);
+    /// ```
+    pub fn new(s: &'a str, window_size: usize) -> Self {
+        assert!(window_size > 0, "Window size must be greater than 0");
+
+        if s.is_empty() {
+            return WindowedWords {
+                s,
+                word_starts: vec![],
+                window_size,
+                current: 0,
+            };
+        }
+
+        // assume first character is not whitespace
+        let mut word_starts = vec![0];
+
+        for (i, _) in s.match_indices(' ') {
```

Comment:
you should be able to store this iterator directly inside `WindowedWords` and pluck from it directly in next. This let's you avoid the heap allocation of the vec

Reply:
pretty sure we will need two. do we want this?